### PR TITLE
Fix #9710 - Autocalculated Cw factor for `ZoneVentilation:WindandStackOpenArea` appears wrong

### LINKED ISSUE #9710: Autocalculated Cw factor for `ZoneVentilation:WindandStackOpenArea` appears wrong
Issue overview
--------------
Originally posted on [Unmet Hours](https://unmethours.com/question/82533/question-about-cw-the-opening-effectiveness-factor-for-zoneventilationwindandstackopenarea/).

ASHRAE HOF:
image

EnergyPlus code:
https://github.com/NREL/EnergyPlus/blob/5b72c372e7302b3d9ab39b73b55071027c6dd392/src/EnergyPlus/ZoneEquipmentManager.cc#L5470-L5473

cc @Nigusse 

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #9710

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc [hunk 1]
```diff
@@ -4696,3 +4687,116 @@ TEST_F(EnergyPlusFixture, ZoneEquipmentManager_SizeZoneEquipment_NoLoadTest)
...
+        //     MCP = AirDensity * CpAir * VVF
+        // <=> Qw = MCP / (AirDensity * CpAir)
+        return MCP / (AirDensity * CpAir);
+    };
+    Real64 areaTimesScheduleValueTimesWind = thisVentilation.OpenArea * 1 * thisZone.WindSpeed;
+    auto calcCw = [&calcQw, &areaTimesScheduleValueTimesWind](Real64 MCP) {
+        auto thisQw = calcQw(MCP);
+        // Qw = Cw * areaTimesScheduleValueTimesWind
+        return thisQw / areaTimesScheduleValueTimesWind;
+    };
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc:4759]: New test.

The function will store the MCP on the Ventilation object, we calculate the Cw from it.

File: tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc [hunk 2]
```diff
@@ -4696,3 +4687,116 @@ TEST_F(EnergyPlusFixture, ZoneEquipmentManager_SizeZoneEquipment_NoLoadTest)
...
+    EXPECT_NEAR(0.55, calcCw(thisVentilation.MCP), 0.0001) << formatFailure();
+
+    // Diagonal winds
+    thisZone.WindDir = thisVentilation.EffAngle + 45;
+    CalcAirFlowSimple(*state);
+    EXPECT_NEAR(0.30, calcCw(thisVentilation.MCP), 0.0001) << formatFailure();
+
+    thisZone.WindDir = thisVentilation.EffAngle - 45;
+    CalcAirFlowSimple(*state);
+    EXPECT_NEAR(0.30, calcCw(thisVentilation.MCP), 0.0001) << formatFailure();
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc:4780]: The angle = 0 works before fix, but the diagonal is wrong.

```shell
[ RUN      ] EnergyPlusFixture.CalcAirFlowSimple_WindAndStackArea

/Users/julien/Software/Others/EnergyPlus2/tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc:4776: Failure
The difference between 0.30 and calcCw(thisVentilation.MCP) is 0.18749999999999994, which exceeds 0.0001, where
0.30 evaluates to 0.29999999999999999,
calcCw(thisVentilation.MCP) evaluates to 0.48749999999999993, and
0.0001 evaluates to 0.0001.
Failed for WindDir=180 and EffAngle=135, absolute angle between opening and wind dir=45

/Users/julien/Software/Others/EnergyPlus2/tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc:4780: Failure
The difference between 0.30 and calcCw(thisVentilation.MCP) is 0.18749999999999994, which exceeds 0.0001, where
0.30 evaluates to 0.29999999999999999,
calcCw(thisVentilation.MCP) evaluates to 0.48749999999999993, and
0.0001 evaluates to 0.0001.
Failed for WindDir=90 and EffAngle=135, absolute angle between opening and wind dir=45
```

File: tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc [hunk 3]
```diff
@@ -4696,3 +4687,116 @@ TEST_F(EnergyPlusFixture, ZoneEquipmentManager_SizeZoneEquipment_NoLoadTest)
...
+        {315, 0.0},  // opposite side
+        {360, 0.0},  // opposite side
+    };
+
+    for (auto &[windDir, expectedCw] : tests) {
+        thisZone.WindDir = windDir;
+
+        CalcAirFlowSimple(*state);
+        EXPECT_NEAR(expectedCw, calcCw(thisVentilation.MCP), 0.0001) << formatFailure();
+    }
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc:4801]: Extend unit testing to test all orientations of wind Dir

File: tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc [hunk 4]
```diff
@@ -4696,3 +4687,116 @@ TEST_F(EnergyPlusFixture, ZoneEquipmentManager_SizeZoneEquipment_NoLoadTest)
...
+
+    state->dataHeatBal->TotVentilation = 1;
+    state->dataHeatBal->AirFlowFlag = true;
+
+    state->dataHeatBal->Ventilation.allocate(1);
+    auto &thisVentilation = state->dataHeatBal->Ventilation(1);
+    // Parameters we really care about
+    thisVentilation.ModelType = DataHeatBalance::VentilationModelType::WindAndStack;
+    thisVentilation.OpenEff = DataGlobalConstants::AutoCalculate;
+    thisVentilation.EffAngle = 135; // Effective angle
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc:4729]: Note this, for the test

File: src/EnergyPlus/ZoneEquipmentManager.cc [hunk 5]
```diff
@@ -5521,10 +5521,60 @@ void CalcAirFlowSimple(EnergyPlusData &state,
...
+                    // | ASHRAE description  | min  | max  | mean | Angle* |
+                    // |---------------------|------|------|------|--------|
+                    // | Perpendicular winds | 0.5  | 0.6  | 0.55 | 0      |
+                    // | Diagonal winds      | 0.25 | 0.35 | 0.3  | 45     |
+                    //
+                    // * Angle is using our convention described above
+                    constexpr Real64 slope = (0.3 - 0.55) / (45 - 0.0);
+                    constexpr Real64 intercept = 0.55;
+                    Cw = intercept + angle * slope;
+                }
```
Comment (jmarrec) [inline src/EnergyPlus/ZoneEquipmentManager.cc:5577]: Fix is here.

I think there's no need to add extra comments on top of these.

**Comment by jmarrec** [comment]
Note that I found another paper will researching this that gives a bit more details than the HoF:

https://www.researchgate.net/profile/Irenilza-Naeaes/publication/234112637_An_algorithm_for_determining_opening_effectiveness_in_natural_ventilation_by_wind/links/54e364e70cf201fe67ec9bf5/An-algorithm-for-determining-opening-effectiveness-in-natural-ventilation-by-wind.pdf

> Nääs (1980) and Bucklin et al. (1991) present the opening effectiveness in more detail and express it as proportional to the wind's angle of incidence:
> * E = 0.60 for wind directions 85 to 90° to openings
> * E = 0.50 for wind directions 70 to 85° to openings
> * E = 0.35 for wind directions 50 to 70° to openings
> * E = 0.30 for wind directions 30 to 50° to openings
> * E = 0.25 for wind directions 10 to 30° to openings

I did **not** change the original  intent of doing a regression on the HoF provided datapoints  (2 of them) because this is a bug fix though.

HoF:
> Cw = effectiness of openings (Cw is assumed to be 0.5 to 0.6 for perpendicular winds and 0.25 to 0.35 for diagonal winds

**Comment by jmarrec** [comment]
Documentation failure has nothing to do with this PR, it's a timeout on GHA when trying to apt update. Lots of comments today a the bottom of https://github.com/actions/runner-images/issues/675

**Edit**: rebooted the run today and it worked

**Comment by Myoldmopar** [comment]
> **Edit**: rebooted the run today and it worked

The docs are building again!?  Woo!!

File: src/EnergyPlus/ZoneEquipmentManager.cc [hunk 6]
```diff
@@ -5521,10 +5521,60 @@ void CalcAirFlowSimple(EnergyPlusData &state,
             if (thisVentilation.OpenEff != DataGlobalConstants::AutoCalculate) {
                 Cw = thisVentilation.OpenEff;
             } else {
-                // linear interpolation between effective angle and wind direction
-                angle = std::abs(WindDirExt - thisVentilation.EffAngle);
-                if (angle > 180.0) angle -= 180.0;
-                Cw = 0.55 + angle / 180.0 * (0.3 - 0.55);
+                //   Wind Dir (β1)                                        90°, min effectiveness
```
Comment (Myoldmopar) [inline src/EnergyPlus/ZoneEquipmentManager.cc:5524]: I feel like making any commentary on this documentation would make me an accidental art critic.  This is impressive.

File: src/EnergyPlus/ZoneEquipmentManager.cc [hunk 7]
```diff
@@ -5521,10 +5521,60 @@ void CalcAirFlowSimple(EnergyPlusData &state,
...
+                //    └────────┘
+                //
+                // * Anything >90 means the wind is blowing in the opposite direction (on the other side), so effectiveness is nil
+                //    ┌──====──┐
+                //    │        │
+                //    └────────┘
+                //       ▲  ▲
+                //       │  │
+
+                angle = 180.0 - std::abs(std::abs(WindDirExt - thisVentilation.EffAngle) - 180);
```
Comment (Myoldmopar) [inline src/EnergyPlus/ZoneEquipmentManager.cc:5560]: This is laid out really well.

File: tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc [hunk 8]
```diff
@@ -4180,7 +4171,7 @@ TEST_F(EnergyPlusFixture, ZoneAirMassFlowBalance_wSourceAndReceivingZone)
     // set zone exhaust nodes to zero and exhaust fan node flow to 1.0 for receiving only zone
     state->dataLoopNodes->Node(state->dataZoneEquip->ZoneEquipConfig(3).ExhaustNode(1)).MassFlowRate = 0.0;
     state->dataLoopNodes->Node(state->dataZoneEquip->ZoneEquipConfig(3).ExhaustNode(2)).MassFlowRate = 1.0;
-    ;
+
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc:4174]: You found all my missing semicolons, thanks!

File: tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc (hunk 3 above)
Comment (Myoldmopar) [reply tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc:4801]: :+1: :+1:

**Review by Myoldmopar — APPROVED** [review]
This looks great!

**Comment by Myoldmopar** [comment]
@jmarrec this looks great, and thanks for the additional research which could be referenced for later enhancement.  This looks ready, just doing final testing locally with develop pulled in.

**Comment by Myoldmopar** [comment]
> > **Edit**: rebooted the run today and it worked
> 
> The docs are building again!? Woo!!

Well...they just failed again for me with the merge from a few minutes ago, so not quite back to normal yet.

**Comment by Myoldmopar** [comment]
Alright everything looks good local.  I'm assuming I'll continue to get the doc failure when I merge, but that's for another day.  Thanks for this fix @jmarrec

**Pull request merged**